### PR TITLE
M4 Report: Working doc

### PR DESCRIPTION
**Pull request by elBeato**
Weitere Kommentare können direkt in den Branch ergänzt werden. 

## Issue:
#192 
#182

### CODE REVIEW

File: doc/m4/FS22-Group-36-M4-Report.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+Then the cards are distributed for the next match. The dealer duty moves one player to the right, the number 
+of cards that is distributed follows the sequence [5,4,3,2,1,2,3,4,5].
+
+For the match with only 1 card there is a special rule. The players don’t see their own card but put their card 
+with the front revealed on their forehead, so only the other players can see the card, but not the player himself. 
+After the 9 matches, the player with the highest score wins.
+
+## 2. Game view
+The order of the screenshots corresponds to the flow of the game, if possible. Matches 4, 3, 2 and the last match 
+are not listed. The only difference to the five-card match is the number of cards.
```
Comment (salwil) [inline doc/m4/FS22-Group-36-M4-Report.md:59]: What do you mean with _Matches 4, 3, 2, are not listed._ ?

File: doc/m4/FS22-Group-36-M4-Report.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+- Weekly meetings of the whole group at the Irchel have simplified communication. Meeting in person has the advantage 
+that problems can be discussed bilaterally and the others can continue to work, but at the same time can listen with 
+one ear.
+	
+- The use of Git and the clarification of how we work with this tool has made it much easier to edit the software 
+efficiently, the main feature of Git we used was: 
+    1. Pull-request
+    2. Feature-Branch 
+
+- As a small team building event, where of course the subject SoPra was extensively discussed, we met for the
```
Comment (salwil) [inline doc/m4/FS22-Group-36-M4-Report.md:90]: haha :+1:

File: doc/m4/FS22-Group-36-M4-Report.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+"Sechseläuten". Coincidentally, there was a team member from the canton of Uri, which was also the guest canton.
```
Comment (salwil) [inline doc/m4/FS22-Group-36-M4-Report.md:91]: And the team member is also a firefighter. He was really impressed and fascinated by the spectacle and the later Cervelat grill on the fire.

File: doc/m4/FS22-Group-36-M4-Report.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+### 4. Conclusion 
+We are extremely satisfied with our performance as a group. We had an ideal mixture of very experienced software 
+developers and those who acquired broad knowledge during the semester. In particular, the people for whom this 
+topic requires extensive experience in the field of programming. We all learned a lot about working together in a
```
Comment (salwil) [inline doc/m4/FS22-Group-36-M4-Report.md:96]: Don't understand this sentence.

File: doc/m4/FS22-Group-36-M4-Report.md (hunk 3 above)
Comment (salwil) [reply doc/m4/FS22-Group-36-M4-Report.md:91]:
```suggestion
"Sechseläuten". Coincidentally, there was a team member from the canton of Uri, which was also the guest canton.
![image](https://user-images.githubusercontent.com/45465496/169947777-4eb655b9-e3f0-4c64-8a7a-de4156249687.png)
```

File: doc/m4/FS22-Group-36-M4-Report.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    2. We test our game online every week
+
+- Weekly meetings of the whole group at the Irchel have simplified communication. Meeting in person has the advantage 
+that problems can be discussed bilaterally and the others can continue to work, but at the same time can listen with 
+one ear.
+	
+- The use of Git and the clarification of how we work with this tool has made it much easier to edit the software 
+efficiently, the main feature of Git we used was: 
+    1. Pull-request
+    2. Feature-Branch
```
Comment (salwil) [inline doc/m4/FS22-Group-36-M4-Report.md:88]:
```suggestion
    2. Feature-Branch 
    3. Github Actions
```

File: doc/m4/FS22-Group-36-M4-Report.md [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+Then one player after the other must announce how many tricks ("Stich", according to https://en.wikipedia.org/wiki/Jass) 
+they will make in this match, 
+starting from the player right from the current dealer. 
+The last one announcing the number of his tricks for this match must announce several tricks such that the total 
+sum of announced tricks is not equal to the number of cards distributed per player in this match.
+
+
+When the trick announcing round is completed, the game starts. The player to the right of the current dealer starts. 
+The suits don’t matter in this game, just the rank, and there is no trump. The player who played the highest 
+rank in a match, i.e. wins the trick and has to deal the first card for the next trick.
```
Comment (salwil) [inline doc/m4/FS22-Group-36-M4-Report.md:37]:
```suggestion
rank in a match, i.e. wins the trick, has to play the first card for the next trick.
```

File: doc/m4/FS22-Group-36-M4-Report.md [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+there are no teams. The game is played with a “Swiss Jass” card set of 36 cards with four suits.
+
+The game is divided into 9 matches (Runden) with changing number of distributed cards per player 
+(5,4,3,2,1,2,3,4,5), and for every match points are distributed.
+
+For the first match, every player gets 5 cards. 
+Then one player after the other must announce how many tricks ("Stich", according to https://en.wikipedia.org/wiki/Jass) 
+they will make in this match, 
+starting from the player right from the current dealer. 
+The last one announcing the number of his tricks for this match must announce several tricks such that the total
```
Comment (salwil) [inline doc/m4/FS22-Group-36-M4-Report.md:31]:
```suggestion
The last one announcing the number of his tricks for this match must announce as many tricks such that the total 
```

File: doc/m4/FS22-Group-36-M4-Report.md [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+many points as possible during the game. Every player plays for himself, 
+there are no teams. The game is played with a “Swiss Jass” card set of 36 cards with four suits.
+
+The game is divided into 9 matches (Runden) with changing number of distributed cards per player 
+(5,4,3,2,1,2,3,4,5), and for every match points are distributed.
+
+For the first match, every player gets 5 cards. 
+Then one player after the other must announce how many tricks ("Stich", according to https://en.wikipedia.org/wiki/Jass) 
+they will make in this match, 
+starting from the player right from the current dealer.
```
Comment (salwil) [inline doc/m4/FS22-Group-36-M4-Report.md:30]:
```suggestion
starting from the player to the right from the current dealer. 
```

File: doc/m4/FS22-Group-36-M4-Report.md [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+Moris Camporesi (19-764-349)
+
+\newpage
+\raggedright
+## 1.  Introduction to the game
+Our game can be played by two up to five people. The goal for every player is to gain as 
+many points as possible during the game. Every player plays for himself, 
+there are no teams. The game is played with a “Swiss Jass” card set of 36 cards with four suits.
+
+The game is divided into 9 matches (Runden) with changing number of distributed cards per player
```
Comment (salwil) [inline doc/m4/FS22-Group-36-M4-Report.md:24]:
```suggestion
The game is divided into 9 matches with changing number of distributed cards per player 
```

Comment (salwil) [reply doc/m4/FS22-Group-36-M4-Report.md:24]: We use this term in the meantime mainly for the round per trick (not for the match anymore).

File: doc/m4/FS22-Group-36-M4-Report.md [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+After the 9 matches, the player with the highest score wins.
+
+## 2. Game view
+The order of the screenshots corresponds to the flow of the game, if possible. Matches 4, 3, 2 and the last match 
+are not listed. The only difference to the five-card match is the number of cards.
+Depends on the result of the game, after the last match, only the winner can see a screw your neighbor cup, 
+well-designed by Carmen. 
+
+
+## 3. Lessen learned
```
Comment (salwil) [inline doc/m4/FS22-Group-36-M4-Report.md:64]:
```suggestion
## 3. Lessons learned
```

File: doc/m4/FS22-Group-36-M4-Report.md [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+#### UI
+
+#### Database
+
+#### Game Logic (side effects)
+
+#### Game Validations
+
+#### Security
+
```
Comment (salwil) [inline doc/m4/FS22-Group-36-M4-Report.md:85]: Not sure, if I understood the term _component_ i.e. the task _Identify your project’s 3-5 main components._ in the right way. Maybe they mean with components also parts of the application that implement a specific logic, e.g. the stacking rule.

File: doc/m4/FS22-Group-36-M4-Report.md [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+#### X.x Spring Data REST
+
+#### X.x Hibernate, JPA
+
+#### X.x Jitsi API
+
+#### X.x Mockito
+
+#### X.x Git
+
```
Comment (salwil) [inline doc/m4/FS22-Group-36-M4-Report.md:73]: Did I forget technologies? Would you remove some of these ones?

File: doc/m4/FS22-Group-36-M4-Report.md [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+with the front revealed on their forehead, so only the other players can see the card, but not the player himself. 
+After the 9 matches, the player with the highest score wins.
+
+## 2. Game view
+The order of the screenshots corresponds to the flow of the game, if possible. Matches 4, 3, 2 and the last match 
+are not listed. The only difference to the five-card match is the number of cards.
+Depends on the result of the game, after the last match, only the winner can see a screw your neighbor cup, 
+well-designed by Carmen. 
+
+### X Technologies
```
[the remark this answers is not in the document]
Comment (salwil) [reply doc/m4/FS22-Group-36-M4-Report.md:63]: Oh, right! Sorry for that.

File: README.md [hunk 14]
```diff
@@ -22,6 +22,64 @@ By group 36 of the course "Software Praktikum"@UZH in FS 22.
...
 
+## Introduction
+World famous card game played with 36 Swiss "Jasskarten". 
+
+## Technologies
+Briefly overview about used main technologies: 
+
+![spring](./doc/img/spring.png)
+![java](./doc/img/gradle.png)
+![rapid api](./doc/img/rapis_api.png)
```
Comment (BacLuc) [inline README.md:33]: rapid api

File: README.md [hunk 15]
```diff
@@ -22,6 +22,64 @@ By group 36 of the course "Software Praktikum"@UZH in FS 22.
...
+
+### Spring Framework
+The Rapid Api Development Architecture provided by spring-data-rest was used to develop the api.
+The PagingAndSortingRepositories provide request handling, deserialization, crud on the database,
+serialization and rendering of errors. Validation was implemented with Bean validations and if
+necessary with Spring event handlers. Side effects were also done with Spring event handlers.
+
+### Hibernate, JPA
+
+### Jitsi API
```
Comment (BacLuc) [inline README.md:48]: i would put that directly in the frontend readme

File: README.md [hunk 16]
```diff
@@ -22,6 +22,64 @@ By group 36 of the course "Software Praktikum"@UZH in FS 22.
...
+(move to FE README)
+
+### Mockito
+
+### Git
+
+
+## High-level components
+### UI
```
Comment (BacLuc) [inline README.md:57]: We have no ui in the backend

File: README.md [hunk 17]
```diff
@@ -22,6 +22,64 @@ By group 36 of the course "Software Praktikum"@UZH in FS 22.
...
+
+### Database
+
+### Game Logic (side effects)
+
+### Game Validations
+
+### Security
+
```
Comment (BacLuc) [inline README.md:66]: What about Serialization/Deserialization?

File: README.md (hunk 16 above)
Comment (elBeato) [reply README.md:57]: I've done it already. Hope this is okey for you @salwil ?

File: README.md [hunk 18]
```diff
@@ -22,6 +22,132 @@ By group 36 of the course "Software Praktikum"@UZH in FS 22.
...
+### Serialization/Deserialization
+Data transmission is done with Jackson JSON serializer and deserializer. Only for transmission of card objects
+the serializer is overwritten by a special serializer (CardSerializer). The reason for that is that displaying
+cards to a user depends on the round. Normally players can see their own cards and cards of the other players
+are hidden. In the round 5 where only one card is distributed per player, it's in reverse. Therefore serialize
+method had to be overwritten such that players can't see their own cards but all the other player's cards instead.
+
+## Launch & Deployment
+@Lucius? :)
```
Comment (salwil) [inline README.md:135]: @BacLuc ?

File: README.md [hunk 19]
```diff
@@ -22,6 +22,132 @@ By group 36 of the course "Software Praktikum"@UZH in FS 22.
...
+annotation is executed after the update of an entity).
+
+### Validations
+Validations are implemented in several places. Format validations are implemented on entities level
+by bean validations. Some Validator classes dedicated to specific entities do further game validations.
+The CardValidator class for example ensures that a player can not play two cards in the same round.
+
+Validation errors as well as technical exceptions from Java or Hibernate are catched and transformed 
+in appropriate HTTP status codes with informal exception messages to prevent technical errors in the
+frontend.
```
Comment (salwil) [inline README.md:100]: Should we create a table with validation errors and respective http status codes?

File: README.md [hunk 20]
```diff
@@ -44,57 +44,43 @@ For building and deploying the software Gradle was in use with the following plu
...
-
-### JPA
-Jakarta Persistence API is the API for storing, retrieving, updating and deleting relational database
-entries in an object oriented context.
-
-### Rapid API
-![rapid api](./doc/img/rapid_api.png)
-
-??? --> Frontend?
+The data was stored in a h2 database. The data is accessed through the Object Relational Mapper hibernate.
```
Comment (BacLuc) [inline README.md:61]: h2 != hibernate

File: README.md [hunk 21]
```diff
@@ -44,57 +44,43 @@ For building and deploying the software Gradle was in use with the following plu
...
-UML diagrams were created with PlantUML plain text language.
-
-### Git, Github
-The versioning of the source code and other files was maintained by Git ont the GitHub platform.
-The GitHubs Actions feature was used for automated build, test and deployment of the software.
+UML diagrams were created with PlantUML.
 
 ## High-level components
 ### Database
-All entities used for the game are persisted during a running game. The main reason is that Spring Data REST
```
Comment (BacLuc) [inline README.md]: The main reason for using a database is that it provides persistence.
The other stuff is done with the magic of spring data rest. But this has nothing to do with the database.

File: README.md [hunk 22]
```diff
@@ -111,57 +97,68 @@ This transferred between frontend and backend by cookies within the http request
 All interactions on entities which are exposed for http requests are protected the WebSecurity configuration
 which only allows to call the /players endpoint without any valid authentication.
 
-| File                                                                                                                                                                                                  | Responsability                                                                                                    |
-|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------|
-| [PlayerEventHandler](https://github.com/sopra-fs22-group-36/screw-your-neighbor-server/blob/main/src/main/java/ch/uzh/ifi/hase/soprafs22/screwyourneighborserver/sideeffects/PlayerEventHandler.java) | Creates security context with authentication token when a new player is created.                                  |
```
Comment (BacLuc) [inline README.md]: You can (and should) use relative URL's in the project README.
Then it links to the file in the branch you opened on github.
And the link also works in your IDE.
And the link is shorter

File: README.md [hunk 23]
```diff
@@ -111,57 +97,68 @@ This transferred between frontend and backend by cookies within the http request
...
 
 To **deploy and release source code to production** the code has to be merged from a feature branch to the main branch. Pushing directly
-to the main branch is not possible because any code being published on the main branch is automatically deployed by a CD pipeline to Heroku.
+to the main branch is disabled by the repository configuration.
 Standard process is to create a pull request which can be merged when at least one other developer gave their approval. Merging is not blocked
 without, but individual deployment without code review and approval is not appreciated by the team.
 
 ## Roadmap
 ### Top features to contribute
-- Sound effects when playing card, winning / loosing trick, winning the game etc.
```
Comment (BacLuc) [inline README.md]: you wouldn't hear it if the server played a sound in the datacenter

File: README.md [hunk 24]
```diff
@@ -22,6 +22,154 @@ By group 36 of the course "Software Praktikum"@UZH in FS 22.
...
+
+#### Spring Data Rest
+The rapid api development architecture provided by spring-data-rest was used to develop the api.
+It provides an easy way to build REST web services on top of the data repositories. The
+PagingAndSortingRepositories and the corresponding entities provide everything for CRUD operations on the database
+(Deserialization, Database Operation, Serialization, Basic Error Handling).
+Validation was implemented with Bean validations and if necessary with Spring event handlers.
+Side effects were also done with Spring event handlers.
+
+### Hibernate
```
Comment (salwil) [inline README.md:58]:
```suggestion
### Hibernate & H2
```

File: README.md (hunk 20 above)
Comment (salwil) [reply README.md:61]:
```suggestion
The data is stored in a H2 database. The data is accessed through the Object Relational Mapper of the Hibernate framework.
```

File: doc/m4/FS22-Group-36-M4-Report.md (hunk 1 above)
Comment (elBeato) [reply doc/m4/FS22-Group-36-M4-Report.md:66]: I mean rounds...thx Sherlok! ;-)

File: doc/m4/FS22-Group-36-M4-Report.md [hunk 25]
```diff
@@ -0,0 +1,168 @@
...
+
+[^1]: Design by Carmen Kirchdorfer @CarmenUZH
+
+## 3. Lessons learned
+We have divided our learnings into three chapters, Challenges, success stories and notes on teamwork.
+
+### 3.1 Challenges
+- The most challenging part was to learn the tech stack in an appropriate time.
+- The next challenge was to review the pull requests in a reasonable amount of time and then to implement the reviews.
+While this was a time-consuming process, it helps to understand the code and how the program works.
```
Comment (salwil) [inline doc/m4/FS22-Group-36-M4-Report.md:112]:
```suggestion
While this was a time-consuming process, it helped to understand the code and how the program work.
```

File: doc/m4/FS22-Group-36-M4-Report.md [hunk 26]
```diff
@@ -0,0 +1,168 @@
...
+Thus, with each pull request, several people knew how far the user stories were implemented.
+- We have mixed feelings about using HATEOAS. On one hand every object had it's
+[IRI](https://en.wikipedia.org/wiki/Internationalized_Resource_Identifier) as identifier, and the relations to other 
+objects was clear. On the other hand the serialization was not consistent with embedded entities, and we had
```
Comment (salwil) [inline doc/m4/FS22-Group-36-M4-Report.md:116]:
```suggestion
objects were clear. On the other hand the serialization was not consistent with embedded entities, and we had
```

File: doc/m4/FS22-Group-36-M4-Report.md [hunk 27]
```diff
@@ -0,0 +1,168 @@
...
+
+![Game playing: one-card-round](Round_with_one_card.png)
+
+### 2.2 Game ending
+At the end of the game it is evaluated who has reached the most points. Different scenarios are visualized in 
+different players.
+
+![End of game: winner page](Winner_screen.png) [^1]
+
+![end of game: looser page](Loser_screen.png)
```
Comment (salwil) [inline doc/m4/FS22-Group-36-M4-Report.md:103]:
```suggestion
![End of game: looser page](Loser_screen.png)
```

File: doc/m4/FS22-Group-36-M4-Report.md [hunk 28]
```diff
@@ -0,0 +1,168 @@
...
+
+![Game playing: stacked round](Stacked_round.png)
+
+![Game playing: one-card-round](Round_with_one_card.png)
+
+### 2.2 Game ending
+At the end of the game it is evaluated who has reached the most points. Different scenarios are visualized in 
+different players.
+
+![End of game: winner page](Winner_screen.png) [^1]
```
Comment (salwil) [inline doc/m4/FS22-Group-36-M4-Report.md:101]: Fusszeile zerstört offenbar die Bildunterschrift. Im PDF hat dieses Bild als einziges keine Bildunterschrift, nur die Referenz zur Fusszeile.

**Pull request merged**